python/dolfinx/plot.py

```diff
@@ -32,39 +33,44 @@ def create_vtk_mesh(mesh: _mesh.Mesh, dim: int, entities=None):
     indices for the associated entry in the mesh geometry.
 
     """
+    cell_type = mesh.topology.cell_type
+    degree = mesh.geometry.cmap.degree
+    if cell_type == _mesh.CellType.prism:
+        raise RuntimeError("Plotting of prism meshes not supported")
+
+    # Use all cells local to process if not supplied
     if entities is None:
         num_cells = mesh.topology.index_map(dim).size_local
         entities = np.arange(num_cells, dtype=np.int32)
     else:
         num_cells = len(entities)
 
-    # Get the indices in the geometry array that correspong to the
-    # topology vertices
-    # NOTE: This linearizes higher order geometries
-    geometry_entities = _cpp.mesh.entities_to_geometry(mesh, dim, entities, False)
+    if dim == mesh.topology.dim:
+        geometry_entities = _cpp.io.extract_vtk_connectivity(mesh)[entities]
+        de = mesh.geometry.cmap.create_dof_layout()
+        num_nodes_per_cell = de.num_dofs
+    else:
+        # NOTE: This linearizes higher order geometries
+        geometry_entities = _cpp.mesh.entities_to_geometry(mesh, dim, entities, False)
+        if degree > 1:
+            warnings.warn("Linearizing topology for higher order sub entities.")
+        e_type = _cpp.mesh.cell_entity_type(cell_type, dim, 0)
 
-    # Array holding the cell type (shape) for each cell
-    cell_type = mesh.topology.cell_type
-    if cell_type == _mesh.CellType.prism:
-        raise RuntimeError("Plotting of prism meshes not supported")
-    degree = mesh.geometry.cmap.degree
-    if degree > 1:
-        warnings.warn("Plotting of higher order mesh topologies is experimental.")
+        # Get cell data and the DOLFINx -> VTK permutation array
+        num_nodes_per_cell = geometry_entities.shape[1]
+        map_vtk = np.argsort(_cpp.io.perm_vtk(e_type, num_nodes_per_cell))
+        geometry_entities = geometry_entities[:, map_vtk]
+
+    # Create mesh topology
+    topology = np.empty((num_cells, num_nodes_per_cell + 1), dtype=np.int32)
+    topology[:, 0] = num_nodes_per_cell
+    topology[:, 1:] = geometry_entities
 
-    e_type = _cpp.mesh.cell_entity_type(cell_type, dim, 0)
+    # Array holding the cell type (shape) for each cell
     vtk_type = _first_order_vtk[cell_type] if degree == 1 else _cpp.io.get_vtk_cell_type(cell_type, mesh.topology.dim)
```

Comment:
Why are both `_first_order_vtk` and  `_cpp.io.get_vtk_cell_type` needed?

Reply:
See the comment at the declaration of the dictionary:
```
# NOTE: This dictionary and following function should be revised when
 # pyvista has better support for arbitrary lagrangian elements, see:
 # https://github.com/pyvista/pyvista/issues/947

 # Cell types can be found at
 # https://vtk.org/doc/nightly/html/vtkCellType_8h_source.html
 ```

Reply:
But the type doesn't fundamentally change the order. `get_vtk_cell_type` effectively give the ID for the cell shape. The 'order' is defined by the number of points per cells.

Reply:
It does not, but it gives nicer rendering of  edges on first order meshes with the traditional cell type than with the arbitrary lagrangian one.